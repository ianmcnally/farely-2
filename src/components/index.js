export { default as Additions } from './additions/additions'
export { default as Footer } from './footer/footer'
export { default as Instructions } from './instructions/instructions'
export { default as MaximumSpend } from './maximum-spend/maximum-spend'
export { default as RemainingBalance } from './remaining-balance/remaining-balance'
