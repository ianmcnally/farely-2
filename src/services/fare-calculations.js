import {
  TRANSACTION_MAX,
  RIDE_COST
} from '../constants'
import range from 'lodash.range'

const getPossibleFaresForSpendingRange = (currentBalance, maximumSpend) => {
  const ridesLeftOnCard = (currentBalance > RIDE_COST) ? Math.floor(currentBalance / RIDE_COST) : 0
  const amountOfRidesPossibleWithinBudget = Math.floor(maximumSpend / RIDE_COST)

  return range(ridesLeftOnCard + 1, amountOfRidesPossibleWithinBudget + 1).map(rideCount => ({
    rides: rideCount,
    fare: rideCount * RIDE_COST
  }))
}

const getCostsForFares = (currentBalance, fares) => fares.map(({ rides, fare }) => {
  const costOfFare = fare - currentBalance
  return { rides, cost: costOfFare }
})

const isDollarAmountAMultipleOfFive = dollarAmount => ((dollarAmount * 100) % 5) === 0

const isValidAmountToAdd = (value, maximumValue) => isDollarAmountAMultipleOfFive(value) && (value <= maximumValue)

export const costForFaresInBudget = (currentBalance = 0, maximumSpend = TRANSACTION_MAX) => {
  const maximumAmountToSpend = (maximumSpend > TRANSACTION_MAX) ? TRANSACTION_MAX : maximumSpend
  const possibleFares = getPossibleFaresForSpendingRange(currentBalance, maximumAmountToSpend)
  const possibleCosts = getCostsForFares(currentBalance, possibleFares)
  const validAmountsToAdd = possibleCosts.filter(({ cost }) => isValidAmountToAdd(cost, maximumAmountToSpend))
  return validAmountsToAdd
}
