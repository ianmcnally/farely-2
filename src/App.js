import React, { Component } from 'react'
import {
  RemainingBalance,
  MaximumSpend,
  Additions,
  Instructions,
  Footer,
} from './components'
import { costForFaresInBudget } from './services/fare-calculations'
import 'normalize.css'
import 'tachyons-borders/css/tachyons-borders.min.css'
import 'tachyons-border-colors/css/tachyons-border-colors.min.css'
import 'tachyons-display/css/tachyons-display.min.css'
import 'tachyons-font-family/css/tachyons-font-family.min.css'
import 'tachyons-font-weight/css/tachyons-font-weight.min.css'
import 'tachyons-forms/css/tachyons-forms.min.css'
import 'tachyons-line-height/css/tachyons-line-height.min.css'
import 'tachyons-max-widths/css/tachyons-max-widths.min.css'
import 'tachyons-spacing/css/tachyons-spacing.min.css'
import 'tachyons-type-scale/css/tachyons-type-scale.min.css'
import 'tachyons-typography/css/tachyons-typography.min.css'
import 'tachyons-utilities/css/tachyons-utilities.min.css'
import 'tachyons-widths/css/tachyons-widths.min.css'
import 'tachyons-lists/css/tachyons-lists.min.css'
import 'tachyons-border-style/css/tachyons-border-style.min.css'
import 'tachyons-border-widths/css/tachyons-border-widths.min.css'

const defaultCurrentBalanceBeforeRender = '0.00'
const defaultCurrentBalanceAfterRender = ''
const defaultMaximumSpend = '40.00'
const defaultFares = undefined

export default class App extends Component {
  state = {
    fares: costForFaresInBudget(
      defaultCurrentBalanceBeforeRender,
      defaultMaximumSpend,
    ),
    currentBalance: defaultCurrentBalanceBeforeRender,
    maximum: defaultMaximumSpend,
  }

  componentDidMount() {
    this.setState({
      currentBalance: defaultCurrentBalanceAfterRender,
      fares: defaultFares,
    })
  }

  handleRemainingChange = (_e, value) => {
    this.setState(currentState => ({
      currentBalance: value,
      fares: costForFaresInBudget(value, currentState.maximum),
    }))
  }

  handleMaximumChange = (_e, value) => {
    this.setState(currentState => ({
      maximum: value,
      fares: costForFaresInBudget(currentState.currentBalance, value),
    }))
  }

  render() {
    const { fares, currentBalance, maximum } = this.state
    const { handleMaximumChange, handleRemainingChange } = this
    const hideInstructions = Boolean(fares)

    return (
      <main className="sans-serif mw5 center">
        <RemainingBalance
          value={currentBalance}
          onChange={handleRemainingChange}
        />
        <MaximumSpend value={maximum} onChange={handleMaximumChange} />
        <Additions fares={fares} />
        <Instructions hide={hideInstructions} />
        <Footer />
      </main>
    )
  }
}
