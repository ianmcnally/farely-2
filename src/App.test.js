import React from 'react';
import App from './App';
import renderShallow from 'render-shallow'
import { findWithType } from 'react-shallow-testutils'
import { Additions, Instructions, RemainingBalance, MaximumSpend } from './components'
import { costForFaresInBudget } from './services/fare-calculations'

describe('when it renders', () => {
  let component

  beforeAll(() => {
    component = renderShallow(<App />).output
  })

  it('renders the App with a zero balance and its fares', () => {
    expect(component).toMatchSnapshot()
  })

})

describe('when it mounts', () => {
  let component

  beforeAll(() => {
    const { instance, rerender } = renderShallow(<App />)

    instance().componentDidMount()

    component = rerender()
  })

  it('renders the App with no current balance and no fares', () => {
    expect(component).toMatchSnapshot()
  })

})

describe('when the remaining balance changes', () => {
  const mockChangeEvent = {}
  const mockChangeValue = '1.00'
  let remainingBalanceComponent

  beforeAll(() => {
    const { output, rerender } = renderShallow(<App />)

    const remainingBalance = findWithType(output, RemainingBalance)
    remainingBalance.props.onChange(mockChangeEvent, mockChangeValue)

    const nextOutput = rerender()
    remainingBalanceComponent = findWithType(nextOutput, RemainingBalance)
  })

  it('updates the value passed to <RemainingBalance>', () => {
    expect(remainingBalanceComponent.props.value).toEqual(mockChangeValue)
  })

})

describe('when the maximum spend changes', () => {
  const mockChangeEvent = {}
  const mockChangeValue = '1.00'
  let maximumSpendComponent

  beforeAll(() => {
    const { output, rerender } = renderShallow(<App />)

    const maximumSpend = findWithType(output, MaximumSpend)
    maximumSpend.props.onChange(mockChangeEvent, mockChangeValue)

    const nextOutput = rerender()
    maximumSpendComponent = findWithType(nextOutput, MaximumSpend)
  })

  it('updates the value passed to <MaximumSpend>', () => {
    expect(maximumSpendComponent.props.value).toEqual(mockChangeValue)
  })

})

describe('when there are a remaining balance and maximum spend values', () => {
  const mockChangeEvent = {}
  const mockBalanceValue = '1.00'
  const mockSpendValue = '40.00'
  let instructionsComponent
  let additionsComponent

  beforeAll(() => {
    const { output, rerender } = renderShallow(<App />)

    const maximumSpend = findWithType(output, MaximumSpend)
    maximumSpend.props.onChange(mockChangeEvent, mockSpendValue)

    const remainingBalance = findWithType(output, RemainingBalance)
    remainingBalance.props.onChange(mockChangeEvent, mockBalanceValue)

    const nextOutput = rerender()
    instructionsComponent = findWithType(nextOutput, Instructions)
    additionsComponent = findWithType(nextOutput, Additions)
  })

  it('sets hide=true on <Instructions>', () => {
    expect(instructionsComponent.props.hide).toBe(true)
  })

  it('passes the fare calculations to <Additions>', () => {
    expect(additionsComponent.props.fares).toEqual(costForFaresInBudget(mockBalanceValue, mockSpendValue))
  })

})
