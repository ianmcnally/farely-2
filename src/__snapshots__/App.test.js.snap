exports[`when it renders renders the App components 1`] = `
<main
  className="sans-serif mw5 center">
  <RemainingBalance
    onChange={[Function]}
    value="" />
  <MaximumSpend
    onChange={[Function]}
    value="40.00" />
  <Additions />
  <Unknown
    hide={false} />
  <Unknown />
</main>
`;
